application/controllers/upload.py

```diff
@@ -108,10 +114,23 @@ def response(status_code, data_group, data_type, payload):
         r.status_code = status_code
     else:
         session['upload_data'] = data
-        r = redirect(url_for('upload_list_data_sets'))
+        r = redirect(redirect_url_for)
 
     return r
 
 
 def json_request(request):
     return request.headers.get('Accept', 'text/html') == 'application/json'
+
+
+def upload_file_and_get_status(data_set):
+    problems, our_problem = \
+        upload_spreadsheet(data_set, request.files['file'])
+
+    if len(problems) == 0:
```

Comment:
This could reuse `get_messages_and_status_for_problems` defined above for error handling.